Lib/multiprocessing/spawn.py

```diff
@@ -59,19 +61,91 @@ def is_forking(argv):
         return False
 
 
+def get_forking_args(argv):
+    '''
+    If the command line indicated we are forking, return (args, kwargs)
+    suitable for passing to spawn_main. Otherwise return None.
+    '''
+    if not is_forking(argv):
+        return None
+
+    args = []
+    kwds = {}
+    for arg in argv[2:]:
+        name, value = arg.split('=')
+        if value == 'None':
+            kwds[name] = None
+        else:
+            kwds[name] = int(value)
+
+    return args, kwds
+
+
+def get_semaphore_tracker_args(argv):
+    '''
+    If the command line indicates we are running the semaphore tracker,
+    return (args, kwargs) suitable for passing to semaphore_tracker.main.
+    Otherwise return None.
+    '''
+    if len(argv) < 2 or argv[1] != '--multiprocessing-semaphore-tracker':
+        return None
+
+    # command ends with main(fd) - extract fd
+    r = int(argv[-1].rsplit('(')[1].split(')')[0])
+
+    args = [r]
+    kwds = {}
+    return args, kwds
+
+
+def get_forkserver_args(argv):
+    '''
+    If the command line indicates we are running the forkserver, return
+    (args, kwargs) suitable for passing to forkserver.main. Otherwise return
+    None.
+    '''
+    if len(argv) < 2 or argv[1] != '--multiprocessing-forkserver':
+        return None
+
+    # command ends with main(listener_fd, alive_r, preload, **kwds) - extract
+    # the args and kwarfs
+    # listener_fd and alive_r are integers
+    # preload is a list
+    # kwds map strings to lists
+    main_args = argv[-1].split('main(')[1].rsplit(')', 1)[0].split(', ', 3)
```

Comment:
Here we are parsing a more difficult expression, which is produced by [forkserver.ForkServer.ensure_running](https://github.com/python/cpython/blob/cb064fc2321ce8673fe365e9ef60445a27657f54/Lib/multiprocessing/forkserver.py#L106-L128). It goes to [forkserver.main](https://github.com/python/cpython/blob/cb064fc2321ce8673fe365e9ef60445a27657f54/Lib/multiprocessing/forkserver.py#L146).

The `listener_fd` and `alive_r` arguments are integers. The `preload` argument is a list, hence I have introduced `ast.literal_eval`.

The keyword arguments come to us as `**{'sys_path': ['list_of_paths'], 'main_path': ['list_of_paths']}`. I used `literal_eval` here as the least awkward option.

Reply:
An alternative to the splitting stuff, since `ast` is already in the mix:

```python
parsed_cmd = ast.parse(argv[-1])
args = [ast.literal_eval(parsed_cmd.body[1].value.args[i]) for i in range(3)]
kwds = ast.literal_eval(parsed_cmd.body[1].value.keywords[0].value)
```